[ROCM] Support TF_ROCM_CLANG for builds with clang host compiler

draganmladjenovic opened the pull request:
Remove TF_NEED_CLANG workarounds

draganmladjenovic committed:
[ROCM] Support TF_ROCM_CLANG for builds with clang host compiler

Remove TF_NEED_CLANG workarounds

i-chaochen commented:
retest Ubuntu-CPU please

jayfurmanek commented:
We'll also need to update the symlink of rocm.bazelrc to point to gpu.bazelrc (instead of gpu_gcc.bazelrc)

jayfurmanek commented:
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/rocm.bazelrc

draganmladjenovic committed:
Point rocm.bazelrc to gpu.bazelrc

draganmladjenovic commented:
Retest Ubuntu-GPU-single please
Retest Ubuntu-CPU please

i-chaochen commented:
I wonder what's situation for this PR? are we capable of switching to clang now?

draganmladjenovic committed:
Use clang for rocm in root .bazelrc

Test doceker doesn't have /userlocal/gpu.bazelrc so flip the switch
in root .bazelrc.

draganmladjenovic commented:
Retest Ubuntu-GPU-multi please.

draganmladjenovic commented:
Retest Ubuntu-CPU please.
Retest Ubuntu-GPU-single please.

draganmladjenovic commented:
@jayfurmanek I think this one is ready. I slightly dislike having to change root .bazelrc in the last commit. Alternative is to create /userlocal/ in Dockerfile.rocm and point bazel in run_gpu_single/multi.sh to use --confing=/userlocal/gpu.bazelrc.

draganmladjenovic commented:
@i-chaochen  Where can I update the docker image for ThirdParty-XLA ci?

i-chaochen commented:
ThirdParty-XLA CI is using the same tensorflow docker image (```rocm/tensorflow-build:latest-python3.9-rocm5.7.0```), may I ask what do you mean you want to update it, and why you want to update it?

draganmladjenovic commented:
> ThirdParty-XLA CI is using the same tensorflow docker image (`rocm/tensorflow-build:latest-python3.9-rocm5.7.0`), may I ask what do you mean you want to update it, and why you want to update it?

It is missing clang-16, so the PR fails, and cannot be merged.

i-chaochen commented:
> rocm/tensorflow-build:latest-python3.9-rocm5.7.0

I think you need to add clang-16 into these three:

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm 

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.rt 

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.ub20
 

```rocm/tensorflow-build:latest-python3.9-rocm5.7.0``` will be updated the latest build.

draganmladjenovic commented:
> > rocm/tensorflow-build:latest-python3.9-rocm5.7.0
> 
> I think you need to add clang-16 into these three:
> 
> https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm
> 
> https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.rt
> 
> https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/Dockerfile.rocm.ub20
> 
> `rocm/tensorflow-build:latest-python3.9-rocm5.7.0` will be updated the latest build.
Ok. I'll split this PR.

draganmladjenovic commented:
> > ThirdParty-XLA CI is using the same tensorflow docker image (`rocm/tensorflow-build:latest-python3.9-rocm5.7.0`), may I ask what do you mean you want to update it, and why you want to update it?
> 
> It is missing clang-16, so the PR fails, and cannot be merged.

My bad. This CI is not blocking the merge. @jayfurmanek The only thing that is missing is the review. Should I include more people into reviewers?

i-chaochen commented on tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc:
```diff
@@ -15,6 +15,21 @@ build --action_env=CACHEBUSTER=565341047
+build:rocm --action_env=CLANG_COMPILER_PATH="/usr/lib/llvm-16/bin/clang"
+# Disable unused-result on rocm builds.
+build:rocm --copt="-Wno-error=unused-result"
+
```
Since we're switching to use gpu.bazelrc, could you also add [xla_cpp_filters](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu_gcc.bazelrc#L184-L187) from gpu_gcc.bazelrc ? Thanks!

jayfurmanek commented on tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc:
```diff
@@ -15,6 +15,21 @@ build --action_env=CACHEBUSTER=565341047
 # Build options for GPU Linux
 build --config=release_gpu_linux
```
This line pulls in the cuda config from the .bazelrc.
We'll need a "release_rocm_linux" config there I think..

jayfurmanek commented on tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc:
```diff
@@ -15,6 +15,21 @@ build --action_env=CACHEBUSTER=565341047
+build:rocm --define=tensorflow_mkldnn_contraction_kernel=0
+build:rocm --repo_env TF_NEED_ROCM=1
+build:rocm --action_env=TF_ROCM_CLANG="1"
+build:rocm --repo_env=BAZEL_COMPILER="/usr/lib/llvm-16/bin/clang"
```
Can we update to clang-17 to match what's upstream now?

jayfurmanek replied:
That will need to be changed in the container as well (merged separately)

draganmladjenovic replied:
> Can we update to clang-17 to match what's upstream now?

Thanks. Missed that. Will update.

jayfurmanek commented on tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc:
```diff
@@ -15,6 +15,21 @@ build --action_env=CACHEBUSTER=565341047
 build --config=release_gpu_linux
 
+# ROCM: Set up compilation ROCM version and paths
+build:rocm --linkopt="-fuse-ld=gold"
```
I think we want to use ldd for the linker

draganmladjenovic replied:
I think I had some issues getting lld on all docker images. I will have to check that.

draganmladjenovic committed:
Move to clang-17 and use lld

jayfurmanek commented:
retest gpu-pycpp please

jayfurmanek commented:
retest gpu-non-pip-multi please

jayfurmanek commented:
retest Ubuntu-GPU-multi please

draganmladjenovic commented:
Retest Ubuntu-GPU-single please.

jayfurmanek reviewed (approved):
Looks good!
Please tag the repo before and after merging:
pre-clang-merge
post-clang-merge
